### ISSUE TITLE
[roq-udp-*] UDP/multicast bridge 

### ISSUE THREAD

**Issue by thraneh** [roq-trading/roq-issues#276]
Gateway-like publisher and subscriber(s).

This will enable wave based message transmission between regions.

**Comment by thraneh** [comment]
Phase 1 version is ready

* `CustomMetrics`
* Prepared for different encoding schemes (FlatBuffers being the default)
* Prepared to support fragments (expecting larger messages when MbP is later being supported)
* Reset based on session
* Sequence number wrap-around
* Drop based on package re-ordering

Missing (not critical for phase 1)

* Proper support for fragments
* Package re-ordering based on sequence number (meaning: don't drop, or wait to restart)

Phase 2 will be focused on adding proper market data support. (Update to follow.)

**Comment by thraneh** [comment]
Phase 2 additions (still work in progress)

* Introduced 2 channels: snapshot and incremental
* All gateways can now optionally support UDP broadcast
* The following events are currently supported:
  * `ReferenceData`
  * `MarketStatus`
  * `TopOfBook`
  * `MarketByPriceUpdate`
  * `TradeSummary`
  * `StatisticsUpdate`
  * `CustomMetricsUpdate`
* The sender can split large payloads into fragments

Missing

* The receiver can **NOT YET** assemble fragmented payloads
* The receiver can **NOT YET** correlate snapshot and incremental

**Comment by thraneh** [comment]
Done

* Message re-assembly from fragments.
* Re-ordering and reset.
* Datagram header now has a space holder for previous sequence number.

Missing

* Correlate snapshot and incremental
* Subscriber to process MbP (it is received and decoded, but not passed through the gateway cache)
* MbP depth limitation (required due to limited total message length
* API to support new update type "unknown"

Notes

We don't have to invalidate all objects when a single datagram was dropped.
All we need to do is mark all objects "unknown" then wait: the next incremental update will tell us (through the previous sequence number) if the dropped datagram relates to it.

The depth of the re-ordering buffer is important since it's directly related to latency when a datagram was dropped.
In other words, we have to wait for the re-ordering buffer to become full before we can detect that the datagram was dropped.

**Comment by thraneh** [comment]
This is now mostly working.

Closing this for now and will open specific issues as they are discovered.
